src/internal/operators/single.ts

```diff
@@ -1,40 +1,77 @@
 import { EmptyError } from '../util/EmptyError';
 
-import { Observer, MonoTypeOperatorFunction, TeardownLogic } from '../types';
+import { MonoTypeOperatorFunction } from '../types';
+import { SequenceError } from '../util/SequenceError';
+import { NotFoundError } from '../util/NotFoundError';
+
+const defaultPredicate = () => true;
 
 /**
- * Returns an Observable that emits the single item emitted by the source Observable that matches a specified
- * predicate, if that Observable emits one such item. If the source Observable emits more than one such item or no
- * items, notify of an IllegalArgumentException or NoSuchElementException respectively. If the source Observable
- * emits items but none match the specified predicate then `undefined` is emitted.
+ * Returns an observable that asserts that only one value is
+ * emitted from the observable that matches the predicate. If no
+ * predicate is provided, then it will assert that the observable
+ * only emits one value.
+ *
+ * In the event that the observable is empty, it will throw an
+ * {@link EmptyError}.
  *
- * <span class="informal">Like {@link first}, but emit with error notification if there is more than one value.</span>
- * ![](single.png)
+ * In the event that two values are found that match the predicate,
+ * or when there are two values emitted and no predicate, it will
+ * throw a {@link SequenceError}
+ *
+ * In the event that no values match the predicate, if one is provided,
+ * it will throw a {@link NotFoundError}
  *
  * ## Example
- * emits 'error'
- * ```ts
- * import { range } from 'rxjs';
- * import { single } from 'rxjs/operators';
  *
- * const numbers = range(1,5).pipe(single());
- * numbers.subscribe(x => console.log('never get called'), e => console.log('error'));
- * // result
- * // 'error'
- * ```
+ * Expect only name beginning with 'B':
  *
- * emits 'undefined'
  * ```ts
- * import { range } from 'rxjs';
+ * import { of } from 'rxjs';
  * import { single } from 'rxjs/operators';
  *
- * const numbers = range(1,5).pipe(single(x => x === 10));
- * numbers.subscribe(x => console.log(x));
- * // result
- * // 'undefined'
+ * const source1 = of(
+ *  { name: 'Ben' },
+ *  { name: 'Tracy' },
+ *  { name: 'Laney' },
+ *  { name: 'Lily' }
+ * );
+ *
+ * source1.pipe(
+ *   single(x => x.name.startsWith('B'))
+ * )
+ * .subscribe(x => console.log(x));
+ * // Emits "Ben"
+ *
+ *
+ * const source2 = of(
+ *  { name: 'Ben' },
+ *  { name: 'Tracy' },
+ *  { name: 'Bradley' },
+ *  { name: 'Lincoln' }
+ * );
+ *
+ * source2.pipe(
+ *   single(x => x.name.startsWith('B'))
+ * )
+ * .subscribe(x => console.log(x));
+ * // Error emitted: SequenceError('Too many values match')
+ *
+ *
+ * const source3 = of(
+ *  { name: 'Laney' },
+ *  { name: 'Tracy' },
+ *  { name: 'Lily' },
+ *  { name: 'Lincoln' }
+ * );
+ *
+ * source3.pipe(
+ *   single(x => x.name.startsWith('B'))
+ * )
+ * .subscribe(x => console.log(x));
+ * // Error emitted: SequenceError('No values match')
```

Comment:
According to the description and the implementation this looks like it should be a `NotFoundError`.